pkg/cmd/root/command.go

```diff
@@ -0,0 +1,132 @@
+	configName       = "gardenctl-v2"
+	targetFilename   = "target.yaml"
+)
+
+var (
+	factory = util.FactoryImpl{}
+)
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	ioStreams := genericclioptions.IOStreams{
+		In:     os.Stdin,
+		Out:    os.Stdout,
+		ErrOut: os.Stderr,
+	}
+
+	rootCmd := &cobra.Command{
+		Use:          "gardenctl",
+		Short:        "gardenctl is a utility to interact with Gardener installations",
+		SilenceUsage: true,
+	}
+
+	rootCmd.AddCommand(target.NewCommand(&factory, target.NewOptions(ioStreams)))
+	rootCmd.AddCommand(version.NewCommand(&factory, version.NewOptions(ioStreams)))
+
+	// Do not precalculate what $HOME is for the help text, because it prevents
+	// usage where the current user has no home directory (which might _just_ be
+	// the reason the user chose to specify an explicit config file).
+	rootCmd.PersistentFlags().StringVar(&factory.ConfigFile, "config", "", fmt.Sprintf("config file (default is $HOME/%s/%s.yaml)", gardenHomeFolder, configName))
+	rootCmd.PersistentFlags().StringVar(&factory.TargetFile, "session", "", fmt.Sprintf("target session file (default is $HOME/%s/%s)", gardenHomeFolder, targetFilename))
+
+	cobra.OnInitialize(initConfig)
+
+	// any error would already be printed, so avoid doing it again here
+	if rootCmd.Execute() != nil {
+		os.Exit(1)
+	}
+}
+
+// initConfig reads in config file and ENV variables if set.
+func initConfig() {
+	var err error
+
+	if factory.ConfigFile != "" {
+		// Use config file from the flag.
+		viper.SetConfigFile(factory.ConfigFile)
+	} else {
+		// Find home directory.
+		home, err := homedir.Dir()
+		cobra.CheckErr(err)
+
+		configPath := filepath.Join(home, gardenHomeFolder)
+
+		// Search config in $HOME/.garden or in path provided with the env variable GCTL_HOME with name ".garden-login" (without extension) or name from env variable GCTL_CONFIG_NAME.
+		envHomeDir, err := homedir.Expand(os.Getenv(envGardenHomeDir))
+		cobra.CheckErr(err)
+
+		viper.AddConfigPath(envHomeDir)
+		viper.AddConfigPath(configPath)
+		if os.Getenv(envConfigName) != "" {
+			viper.SetConfigName(os.Getenv(envConfigName))
+		} else {
+			viper.SetConfigName(configName)
+		}
+	}
+
+	viper.SetEnvPrefix(envPrefix)
+	viper.AutomaticEnv() // read in environment variables that match
+
+	// If a config file is found, read it in.
+	if err := viper.ReadInConfig(); err != nil {
+		klog.Errorf("failed to read config file: %v", err)
+	}
+
+	// initialize the factory
+
+	// prefer an explicit GCTL_HOME env,
+	// but fallback to the system-defined home directory
+	home := os.Getenv(envGardenHomeDir)
+	if len(home) == 0 {
+		home, err = homedir.Dir()
+		cobra.CheckErr(err)
+
+		home = filepath.Join(home, gardenHomeFolder)
+	}
+
+	// prefer -session, then an explicit GCTL_HOME env,
+	// but fallback to the system-defined home directory
```

Comment:
> prefer -session

do not understand

Reply:
Yeah I've removed the notion of "sessions" again now. The comment was .. inspired, yet misleading. I am not actually sure how we want to allow users to switch between targets (we probably want to offer that, but _how_ exactly?), so it's probably best if we do it in another PR.